Add AnchorStateRegistry Implementation to OPSM

maurelian opened the pull request:
### TL;DR

Added support for starting anchor roots in the OPStackManager deployment process.

### What changed?

- Updated `DeployOPChainInput` to include `startingAnchorRoots` field
- Modified `OPStackManager` to accept and use `startingAnchorRoots` in the deployment process
- Updated `DeployOPChain` script to pass `startingAnchorRoots` to the `OPStackManager`
- Added initialization for `AnchorStateRegistry` in the deployment process
- Updated relevant tests to include `startingAnchorRoots` in their setup and assertions

maurelian committed:
feat: Add ASR proxy to OPStackManager

maurelian committed:
feat: Add AnchorStateRegistry blueprint

maurelian committed:
feat: Add AnchorStateRegistry Implementation

maurelian commented:
* **#11955** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/11955?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#11952** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/11952?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 1 other dependent PR ([#11953](https://github.com/ethereum-optimism/optimism/pull/11953) <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/11953?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>)
* `develop`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @maurelian and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

mds1 commented on packages/contracts-bedrock/scripts/DeployOPChain.s.sol:
```diff
@@ -70,6 +71,17 @@ contract DeployOPChainInput is BaseDeployIO {
         }
     }
 
+    function set(bytes4 _sel, AnchorStateRegistry.StartingAnchorRoot[] memory _roots) public {
```
If we're only deploying straight to permissioned games for now (which I believe we are, cause of the cyclic dependency), does that help us cut any scope by not needing to add support and tests for setting initial anchor roots? If including them is a low lift though, we may as well do it

maurelian replied:
I don't think the lift is huge, do you see a much missing in the way of testing here? 

Before cutting it, we'd need to discuss with Proofs. My impression is that empty anchor roots is a possibility but not necessarily the default with the permissioned game.

mds1 replied:
Ok, just checked where `startingAnchorRoots` is used and you're right the current testing is sufficient. Just wanted to raise this as an option but it seems unnecessary to cut this scope, so I'll resolve this

mds1 commented on packages/contracts-bedrock/test/DeployOPChain.t.sol:
```diff
@@ -31,6 +31,8 @@ import { L1ERC721Bridge } from "src/L1/L1ERC721Bridge.sol";
 import { L1StandardBridge } from "src/L1/L1StandardBridge.sol";
 import { OptimismMintableERC20Factory } from "src/universal/OptimismMintableERC20Factory.sol";
 
+import "src/dispute/lib/Types.sol";
```
nit: I know this is usually imported this way in other files, but let's stick with named imports if we can for clarity
```suggestion
import { ... } from "src/dispute/lib/Types.sol";
```

maurelian replied:
Done

mds1 commented on packages/contracts-bedrock/test/DeployOPChain.t.sol:
```diff
@@ -336,9 +338,18 @@ contract DeployOPChain_TestBase is Test {
+        startingAnchorRoots.push(
+            AnchorStateRegistry.StartingAnchorRoot({
+                gameType: GameTypes.CANNON,
+                outputRoot: OutputRoot({ root: Hash.wrap(keccak256("defaultOutputRoot")), l2BlockNumber: 400 })
```
This l2BlockNumber is arbitrary right?

maurelian replied:
Yes, it's a continuation of the numbering above.

mds1 commented on packages/contracts-bedrock/test/DeployOPChain.t.sol:
```diff
@@ -389,7 +400,7 @@ contract DeployOPChain_Test is DeployOPChain_TestBase {
     }
 
-    function testFuzz_run_memory_succeeds(bytes32 _seed) public {
+    function testFuzz_run_memory_succeedsX(bytes32 _seed) public {
```
typo?
```suggestion
    function testFuzz_run_memory_succeeds(bytes32 _seed) public {
```

maurelian replied:
yeah. 
Helped me to isolate the test.

mds1 commented on packages/contracts-bedrock/test/DeployOPChain.t.sol:
```diff
@@ -399,6 +410,15 @@ contract DeployOPChain_Test is DeployOPChain_TestBase {
         basefeeScalar = uint32(uint256(hash(_seed, 6)));
         blobBaseFeeScalar = uint32(uint256(hash(_seed, 7)));
         l2ChainId = uint256(uint256(hash(_seed, 8)));
+        uint256 numStartingAnchorRoots = bound(uint256(uint256(hash(_seed, 9))), 1, 10);
```
Can we add a comment with rationale for these bounds? e.g. why not a lower bound of 0, and what's the max upper bound we might see in practice?

maurelian replied:
Change this to add one root per game type per expected usage.

mds1 commented on packages/contracts-bedrock/test/DeployOPChain.t.sol:
```diff
@@ -399,6 +410,15 @@ contract DeployOPChain_Test is DeployOPChain_TestBase {
+            startingAnchorRoots.push(
+                AnchorStateRegistry.StartingAnchorRoot({
+                    gameType: GameTypes.CANNON,
+                    outputRoot: OutputRoot({ root: Hash.wrap(keccak256(abi.encode(_seed, i))), l2BlockNumber: 400 + i })
```
For my own understanding: is having each root incremented by 1 block a requirement, or it's arbitrary and we just increment by 1 each loop iteration for testing?

maurelian replied:
It was arbitrary, but I've improved this section to be more true to real world usage.

maurelian committed:
feat: Return startingAnchorInputs as bytes

The op-deployer tooling does not support structs, therefore we need to
return a more generic type for compatibility.

maurelian committed:
rebuild snapshots

maurelian committed:
Merge branch 'develop' into opsm/fpac-asr-imp

maurelian committed:
fix: ASR initializer encoding

maurelian committed:
Merge branch 'develop' into opsm/fpac-asr-imp

mds1 commented on packages/contracts-bedrock/src/L1/OPStackManager.sol:
```diff
@@ -55,6 +55,9 @@ contract OPStackManager is ISemver, Initializable {
         uint32 basefeeScalar;
         uint32 blobBasefeeScalar;
         uint256 l2ChainId;
+        // The correct type is AnchorStateRegistry.StartingAnchorRoot[] memory,
+        // but OP Deployer does not yet support structs.
+        bytes startingAnchorRoots;
```
Do you think keeping this as `bytes` is ok, or should this be a TODO?

maurelian replied:
Unless it becomes a headache, I think it's OK to live with this. So far it doesn't seem to affect the external interface too much.

mds1 commented on packages/contracts-bedrock/scripts/DeployOPChain.s.sol:
We should add `assertValidAnchorStateRegistryProxy()` and `assertValidAnchorStateRegistryImpl()` method also

mds1 reviewed (approved):
LGTM pending assertions

Inphi reviewed (approved):
I don't have enough context to review everything in the PR. But can confirm the ASR initialization looks good.

maurelian committed:
handoff commit with op-deployer debugging

debugging op-deployer test
wip: literal anchor roots

mslipper commented on op-chain-ops/deployer/pipeline/opchain.go:
```diff
@@ -62,6 +62,7 @@ func DeployOPChain(ctx context.Context, env *Env, intent *state.Intent, st *stat
 						BasefeeScalar:          1368,
 						BlobBaseFeeScalar:      801949,
 						L2ChainId:              chainID.Big(),
+						StartingAnchorRoots:    []byte{},
```
Is this value constant? If so, we should hardcode it rather than requiring is to be in the intent. If it's in the intent users will have to specify it prior to deploying every chain.

mds1 replied:
I think for permissioned chains it can be an arbitrary constant, based on [this convo](https://discord.com/channels/1244729134312198194/1273519241076670526/1285673518494650482), but should confirm with @Inphi 
<img width="900" alt="image" src="https://github.com/user-attachments/assets/89fc74c2-8642-4eb0-98f4-b9f9b17e0d40">
<img width="878" alt="image" src="https://github.com/user-attachments/assets/62fb28cd-f666-4553-946b-00d95a4c2987">

maurelian replied:
This value is an array of `struct AnchorRoot { GameType, OutputRoot}`, so the _correct_ value will be different for each chain, but because the anchor state is part of the [cyclic dependency](https://github.com/ethereum-optimism/design-docs/pull/79), we won't actually know what the correct value is when deploying a new OP Chain. 

However it must be non-zero because per @Inphi in discord (I've replaced the link he gave with a permalink):

> Actually the starting anchor should be non-zero. Otherwise, the permissioned game will fail this [check](https://github.com/ethereum-optimism/optimism/blob/c21d0c5fa04faa116ef9572f5f9b36d3e03f3a36/packages/contracts-bedrock/src/dispute/FaultDisputeGame.sol). 

So, in order to get this over the line, we may need to simply hardcode some nonsense initial output roots (ie. `0xdead....dead`). That's presumably safe as no meaningful proof would be likely to be discovered which enables a fraudulent withdrawal.

mslipper committed:
test and golang fixes

mslipper reviewed (approved):
Prior to merging, can we refactor the anchor state for these changes to live outside the intent? Or, alternatively, can you have the `init` command add it to the intent automatically?

Edit: pushed a commit that does this.

mslipper committed:
hardcode permissioned state

maurelian commented on packages/contracts-bedrock/scripts/DeployOPChain.s.sol:
```diff
@@ -70,6 +71,14 @@ contract DeployOPChainInput is BaseDeployIO {
         }
     }
 
+    function set(bytes4 _sel, bytes memory _value) public {
```
Just for my own edification, why was it necessary to remove the struct typing on the argument?

mslipper replied:
The internal struct fields need to be primitive types so that the Go VM engine can generate getter/setter methods for them. Since the internal field is no longer the anchor state struct, I updated this `set` method to take regular `bytes`.

mds1 committed:
hardcode 0xdead as the starting anchor root

mds1 committed:
Merge branch 'develop' into opsm/fpac-asr-imp

mds1 committed:
chore: fix semver lock

mds1 committed:
fix: no permissionless root, remove hash from 0xdead

Inphi commented on op-chain-ops/deployer/pipeline/opchain.go:
```diff
@@ -13,10 +13,10 @@ import (
+// and no root for the permissionless game. It was generated with a console.log in DeployOpChain.s.sol
 var PermissionedGameStartingAnchorRoots = []byte{
-	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x20, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x00, 0x00, 0x ...
+	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x20, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x ...
```
@mds1 Is this the abi encoded anchor roots array? It doesn't match what I expect for a 1-sized 0xdead root?

mds1 replied:
Fixed in ca8be30

mds1 committed:
fix: use 0xdead root properly

mds1 committed:
fix: set the override in the input contract

mds1 committed:
Merge branch 'develop' into opsm/fpac-asr-imp

mslipper committed:
Merge remote-tracking branch 'upstream/develop' into opsm/fpac-asr-imp

mslipper committed:
Fix tests and accidental mutation of `Implementation` struct

mslipper commented on packages/contracts-bedrock/src/L1/OPStackManager.sol:
```diff
@@ -345,9 +352,14 @@ contract OPStackManager is ISemver, Initializable {
-    /// @notice Returns the implementation data for a contract name.
-    function getLatestImplementation(string memory _name) internal view returns (Implementation storage) {
-        return implementations[latestRelease][_name];
+    /// @notice Returns the implementation data for a contract name. Makes a copy of the internal
```
@maurelian @mds1 - I updated this method to return a copy of the underlying data in storage. The code [here](https://github.com/ethereum-optimism/optimism/pull/11955/files#diff-adc000b158932be783e0ea3e4fac762b4aa8a6310d4088675627630dc977eaddR302-R306) was overwriting the DisputeGameFactory's release info in the `implementations` map, which was causing subsequent deployments using the OPSM to fail.

mds1 replied:
Nice catch! That makes sense. This fix is ok to unblock this PR, in a future PR I'd like to change back from `memory` to `storage` (it will be cheaper and smaller code), then we just don't assign to `impl` for the ASR logic there

mslipper committed:
lint

mslipper committed:
semver

maurelian commented on op-chain-ops/deployer/opsm/opchain.go:
```diff
@@ -9,6 +9,12 @@ import (
 )
 
+// PermissionedGameStartingAnchorRoots is a root of bytes32(hex"dead") for the permissioned game at block 0,
+// and no root for the permissionless game. It was generated with a console.log in DeployOpChain.s.sol
```
I don't think we need to keep this detail around. 

```suggestion
// and no root for the permissionless game.
```

mds1 committed:
Update op-chain-ops/deployer/opsm/opchain.go

Co-authored-by: Maurelian <john@oplabs.co>

mds1 commented on packages/contracts-bedrock/scripts/DeployOPChain.s.sol:
```diff
@@ -116,6 +118,26 @@ contract DeployOPChainInput is BaseDeployIO {
+        // anchor root and deploy a new permissioned dispute game contract anyway.
+        //
+        // You can `console.logBytes(abi.encode(defaultStartingAnchorRoots))` to get the bytes that
+        // are hardcoded into `op-chain-ops/deployer/pipeline/opchain.go`
```
```suggestion
        // are hardcoded into `op-chain-ops/deployer/opsm/opchain.go`
```

mds1 committed:
Update packages/contracts-bedrock/scripts/DeployOPChain.s.sol
